error during artifact download: landlock failed to lock (NixOS)

### Nomad version
Output from `nomad version`

1.6.1

### Operating system and Environment details

Nixos 23.05 with additional package nomad 1.6.1
```
# uname --all                                                                                                                                                                                                                                                     
Linux cosmo 6.1.53 #1-NixOS SMP PREEMPT_DYNAMIC Wed Sep 13 07:43:05 UTC 2023 x86_64 GNU/Linux
```

### Issue

Downloading artifacts fails the task with some landlock error

Randomly guessing: go-getter or Nomad require some binary dependency not in the PATH of Nomad (can happen with NixOS since there are less defaults for installed binaries)

### Reproduction steps

Simply using `artifact {` and running the job

#### Expected Result

Artifact downloaded as usual

#### Actual Result

Task failed and the logs above

### Job file (if appropriate)

```hcl
job "prometheus" {
  datacenters = ["dc1"]
  type        = "service"

  group "monitoring" {
    count = 1

    network {
      mode = "bridge"
      dns {
        servers = [...]
      }
      port "http" {}
    }

    constraint {
      attribute = "${attr.nomad.bridge.hairpin_mode}"
      value     = true
    }

    restart {
      attempts = 2
      interval = "30m"
      delay    = "15s"
      mode     = "fail"
    }

    ephemeral_disk {
      size    = 256 # MB
      migrate = true
      sticky  = true
    }

    task "prometheus" {
      template {
        change_mode = "restart"
        destination = "local/prometheus.yml"

        data = <<EOH
---
global:
  scrape_interval:     30s
  evaluation_interval: 30s

rule_files:
  - /local/mimir_rules.yaml


scrape_configs:
  # omitted for brevity

EOH
      }
      artifact {
        source      = "https://github.com/grafana/mimir/raw/main/operations/mimir-mixin-compiled/rules.yaml"
        mode        = "file"
        destination = "local/mimir_rules.yaml"
      }


      driver = "docker"

      config {
        image = "prom/prometheus:latest"

        volumes = [
          "local/prometheus.yml:/etc/prometheus/prometheus.yml",
        ]

        args = [
          "--web.route-prefix=/",
          "--web.external-url=http://prometheus.traefik",
          "--config.file=/etc/prometheus/prometheus.yml",
          "--enable-feature=agent",
          "--web.enable-remote-write-receiver",
          "--enable-feature=exemplar-storage"
        ]

        ports = ["http"]
      }

      service {
        name     = "prometheus"
        provider = "nomad"
        port     = "http"

        check {
          name     = "alive"
          type     = "tcp"
          interval = "10s"
          timeout  = "2s"
        }
      }
    }
  }
}

```

<!--
If possible please post relevant logs in the issue.
`
Logs and other artifacts may also be sent to: nomad-oss-debug@hashicorp.com

Please link to your Github issue in the email and reference it in the subject
line:

> To: nomad-oss-debug@hashicorp.com
>
> Subject: GH-1234: Errors garbage collecting allocs

Emails sent to that address are readable by all HashiCorp employees but are
*not* publicly visible.
-->

### Nomad Server logs (if appropriate)

### Nomad Client logs (if appropriate)


```
Oct 11 11:47:04 cosmo nomad[4074065]:     2023-10-11T11:47:04.677+0100 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=4bd91cd4-0560-4e4a-a506-af97f985c137 task=prometheus type="Downloading Artifacts" msg="Client is downloading artifacts" failed=false
Oct 11 11:47:04 cosmo nomad[4074065]:     2023-10-11T11:47:04.728+0100 [ERROR] client.artifact: sub-process: OUTPUT="failed to sandbox artifact-isolation process: landlock failed to lock: no such file or directory"
Oct 11 11:47:04 cosmo nomad[4074065]:     2023-10-11T11:47:04.729+0100 [INFO]  client.alloc_runner.task_runner: Task event: alloc_id=4bd91cd4-0560-4e4a-a506-af97f985c137 task=prometheus type="Failed Artifact Download" msg="failed to download artifact \"https://github.com/grafana/mimir/raw/main/operations/mimir-mixin-compiled/rules.yaml\": getter subprocess failed: exit status 1" failed=false
Oct 11 11:47:04 cosmo nomad[4074065]:     2023-10-11T11:47:04.730+0100 [ERROR] client.alloc_runner.task_runner: prestart failed: alloc_id=4bd91cd4-0560-4e4a-a506-af97f985c137 task=prometheus error="prestart hook \"artifacts\" failed: failed to download artifact \"https://github.com/grafana/mimir/raw/main/operations/mimir-mixin-compiled/rules.yaml\": getter subprocess failed: exit status 1"
```

--- a/.changelog/27149.txt
+++ b/.changelog/27149.txt
@@ -1,0 +1,3 @@
+```release-note:improvement
+landlock: check paths exist on setup
+```

--- a/client/allocrunner/taskrunner/getter/util_default.go
+++ b/client/allocrunner/taskrunner/getter/util_default.go
@@ -7,6 +7,8 @@ package getter
 
 import (
 	"path/filepath"
+
+	log "github.com/hashicorp/go-hclog"
 )
 
 // lockdown is not available by default
@@ -15,7 +17,7 @@ func lockdownAvailable() bool {
 }
 
 // lockdown is not implemented by default
-func lockdown(string, string, []string) error {
+func lockdown(log.Logger, string, string, []string) error {
 	return nil
 }
 

--- a/client/allocrunner/taskrunner/getter/util_linux.go
+++ b/client/allocrunner/taskrunner/getter/util_linux.go
@@ -6,13 +6,23 @@
 package getter
 
 import (
+	"errors"
 	"os"
 	"path/filepath"
 
+	log "github.com/hashicorp/go-hclog"
 	"github.com/mitchellh/go-homedir"
 	"github.com/shoenig/go-landlock"
 	"golang.org/x/sys/unix"
 )
+
+// initialDirs are the initial set of paths configured for landlock
+var initialDirs = map[string]string{
+	"/bin":           "rx",
+	"/usr/bin":       "rx",
+	"/usr/local/bin": "rx",
+	"/usr/libexec":   "rx",
+}
 
 // findHomeDir returns the home directory as provided by os.UserHomeDir. In case
 // os.UserHomeDir returns an error, we return /root if the current process is being
@@ -57,7 +67,7 @@ func lockdownAvailable() bool {
 // dir - the task directory
 //
 // Only applies to Linux, when available.
-func lockdown(allocDir, taskDir string, extra []string) error {
+func lockdown(l log.Logger, allocDir, taskDir string, extra []string) error {
 	// landlock not present in the kernel, do not sandbox
 	if !landlock.Available() {
 		return nil
@@ -66,12 +76,26 @@ func lockdown(allocDir, taskDir string, extra []string) error {
 		landlock.DNS(),
 		landlock.Certs(),
 		landlock.Shared(),
-		landlock.Dir("/bin", "rx"),
-		landlock.Dir("/usr/bin", "rx"),
-		landlock.Dir("/usr/local/bin", "rx"),
-		landlock.Dir("/usr/libexec", "rx"),
 		landlock.Dir(allocDir, "rwc"),
 		landlock.Dir(taskDir, "rwc"),
+	}
+
+	// Add the initial directories
+	for p, mode := range initialDirs {
+		_, err := os.Stat(p)
+		if err != nil {
+			if errors.Is(err, os.ErrNotExist) {
+				// paths that do not exist are skipped
+				l.Debug("landlock is skipping path that does not exist", "path", p)
+			} else {
+				// other errors should be logged to provide context on why
+				// the path is not included
+				l.Warn("landlock setup failed to stat path, skipping", "path", p, "error", err)
+			}
+
+			continue
+		}
+		paths = append(paths, landlock.Dir(p, mode))
 	}
 
 	for _, p := range extra {

--- a/client/allocrunner/taskrunner/getter/util_windows.go
+++ b/client/allocrunner/taskrunner/getter/util_windows.go
@@ -8,6 +8,8 @@ package getter
 import (
 	"os"
 	"path/filepath"
+
+	log "github.com/hashicorp/go-hclog"
 )
 
 // lockdown is not implemented on Windows
@@ -16,7 +18,7 @@ func lockdownAvailable() bool {
 }
 
 // lockdown is not implemented on Windows
-func lockdown(string, string, []string) error {
+func lockdown(log.Logger, string, string, []string) error {
 	return nil
 }
 

--- a/client/allocrunner/taskrunner/getter/z_getter_cmd.go
+++ b/client/allocrunner/taskrunner/getter/z_getter_cmd.go
@@ -6,6 +6,7 @@ package getter
 import (
 	"os"
 
+	log "github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/nomad/helper/subproc"
 )
 
@@ -17,6 +18,11 @@ const (
 
 func init() {
 	subproc.Do(SubCommand, func() int {
+		l := log.New(&log.LoggerOptions{
+			Output:      os.Stderr,
+			DisableTime: true,
+			Level:       log.Debug,
+		})
 
 		// get client and artifact configuration from standard IO
 		env := new(parameters)
@@ -34,7 +40,7 @@ func init() {
 
 		// sandbox the host filesystem for this process
 		if !env.DisableFilesystemIsolation {
-			if err := lockdown(env.AllocDir, env.TaskDir, env.FilesystemIsolationExtraPaths); err != nil {
+			if err := lockdown(l, env.AllocDir, env.TaskDir, env.FilesystemIsolationExtraPaths); err != nil {
 				subproc.Print("failed to sandbox %s process: %v", SubCommand, err)
 				return subproc.ExitFailure
 			}

